### PR TITLE
Added User Assignable Roles

### PR DESCRIPTION
**Pull request by iShauny**
Users can now assign roles themselves
+ iam to add a role, >iamnot to remove a role
+ iamconfig add/remove to modify the assignable roles

### CODE REVIEW

File: Floofbot/Modules/UserAssignableRoleCommands.cs [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+            // checks if a word exists in the filter db
+            bool isRoleInDb = _floofDb.UserAssignableRoles.AsQueryable().Where(r => r.ServerId == guild.Id).Where(r => r.RoleId == role.Id).Any();
+            return isRoleInDb;
+        }
+
+        [Summary("Join a role that the server administrators allow you to")]
+        [Command("iam")]
+        public async Task IAm([Summary("role name")][Remainder]string roleName)
+        {
+            if (roleName == null)
```
Comment (JasperGray) [inline Floofbot/Modules/UserAssignableRoleCommands.cs:45]: I think you might need to change the line above to `[Remainder]string roleName=null` for this to work, or maybe it has to use the empty string `""` as the default parameter instead.

File: Floofbot/Modules/UserAssignableRoleCommands.cs [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+            // find the role by iterating over guild roles
+            SocketRole role = GetRole(roleName.ToLower(), Context.Guild);
+            if (role == null)
+            {
+                await Context.Channel.SendMessageAsync("Unable to find a role with that name.");
+                return;
+            }
+
+            // check that they are actually allowed to join the role
+            bool roleExistsInDb = CheckRoleEntryExists(role, Context.Guild);
```
Comment (JasperGray) [inline Floofbot/Modules/UserAssignableRoleCommands.cs:60]: So we just have to make sure the roles in the table are all user-assignable ones?

File: Floofbot/Modules/UserAssignableRoleCommands.cs [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+            catch (Exception ex)
+            {
+                await Context.Channel.SendMessageAsync("Unable to give you that role. I may not have the permissions.");
+                Log.Error("Error trying to add a role onto a user: " + ex);
+                return;
+            }
+
+        }
+
+        [Summary("Remove a role that the server administrators allow you to")]
```
Comment (JasperGray) [inline Floofbot/Modules/UserAssignableRoleCommands.cs:93]: Can shorten to "Remove a role"

File: Floofbot/Modules/UserAssignableRoleCommands.cs [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+        [Command("iamnot")]
+        public async Task IAmNot([Summary("role name")][Remainder]string roleName)
+        {
+            if (roleName == null)
```
Comment (JasperGray) [inline Floofbot/Modules/UserAssignableRoleCommands.cs:97]: Same as above: does the usage message ever show?

File: Floofbot/Modules/UserAssignableRoleCommands.cs [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+
+            // find the role by iterating over guild roles
+            SocketRole role = GetRole(roleName.ToLower(), Context.Guild);
+            if (role == null)
+            {
+                await Context.Channel.SendMessageAsync("Unable to find a role with that name.");
+                return;
+            }
+
+            // check that they are actually allowed to join the role
```
Comment (JasperGray) [inline Floofbot/Modules/UserAssignableRoleCommands.cs:111]: comment is a bit misleading

File: Floofbot/Modules/UserAssignableRoleCommands.cs [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+                    try
+                    {
+                        await usr.RemoveRoleAsync(role, new RequestOptions { AuditLogReason = "User removed the role themselves." });
+                        await Context.Channel.SendMessageAsync($"You no longer have the '{role.Name}' role.");
+                        return;
+                    }
+                    catch (Exception ex)
+                    {
+                        await Context.Channel.SendMessageAsync("Unable to remove that role. I may not have the permissions.");
+                        Log.Error("Error trying to add a role onto a user: " + ex);
```
Comment (JasperGray) [inline Floofbot/Modules/UserAssignableRoleCommands.cs:134]: remove*

File: Floofbot/Modules/UserAssignableRoleCommands.cs [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+            return isRoleInDb;
+        }
+
+        [Summary("Configure a role as joinable")]
+        [Command("add")]
+        public async Task AddRole([Summary("role name")][Remainder]string roleName)
+        {
+            if (roleName == null)
+            {
+                await Context.Channel.SendMessageAsync("", false, new EmbedBuilder { Description = $"💾 Usage: `iam add [rolename]`", Color = EMBED_COLOUR }.Build());
```
Comment (JasperGray) [inline Floofbot/Modules/UserAssignableRoleCommands.cs:177]: Usage should be .iamconfig

File: Floofbot/Modules/UserAssignableRoleCommands.cs [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+            }
+        }
+
+        [Summary("Configure a role to no longer be joinable")]
+        [Command("remove")]
+        public async Task RemoveRole([Summary("role name")][Remainder]string roleName)
+        {
+            if (roleName == null)
+            {
+                await Context.Channel.SendMessageAsync("", false, new EmbedBuilder { Description = $"💾 Usage: `iam remove [rolename]`", Color = EMBED_COLOUR }.Build());
```
Comment (JasperGray) [inline Floofbot/Modules/UserAssignableRoleCommands.cs:221]: .iamconfig

File: Floofbot/Modules/UserAssignableRoleCommands.cs [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+            }
+            // user does not have role
+            await Context.Channel.SendMessageAsync("You do not have that role.");
+        }
+    }
+
+    [Summary("Configures the joinable user roles")]
+    [RequireUserPermission(GuildPermission.Administrator)]
+    [Group("iamconfig")]
+    public class UserAssignableRoleAdminCommands : InteractiveBase
```
[the remark this answers is not in the document]
Comment (iShauny) [reply Floofbot/Modules/UserAssignableRoleCommands.cs:147]: Another class within the same name space

File: Floofbot/Services/Repository/Models/UserAssignableRole.cs [hunk 10]
```diff
@@ -0,0 +1,12 @@
+﻿using System.ComponentModel.DataAnnotations;
+
+namespace Floofbot.Services.Repository.Models
+{
+    public partial class UserAssignableRole
+    {
+        [Key]
```
[the remark this answers is not in the document]
Comment (iShauny) [reply Floofbot/Services/Repository/Models/UserAssignableRole.cs:7]: Can do that, but role id is unique anyway

File: Floofbot/Modules/UserAssignableRoleCommands.cs [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+using Discord.WebSocket;
+using Serilog;
+
+namespace Floofbot.Modules
+{
+    [Summary("Join roles that the administrator have allowed you to")]
+    public class UserAssignableRoleCommands : InteractiveBase
+    {
+        private FloofDataContext _floofDb;
+        private readonly Discord.Color EMBED_COLOUR = new Discord.Color((uint)new Random().Next(0x1000000));
```
[the remark this answers is not in the document]
Comment (iShauny) [reply Floofbot/Modules/UserAssignableRoleCommands.cs:19]: Created a method to handle this

File: Floofbot/Modules/UserAssignableRoleCommands.cs [hunk 12]
```diff
@@ -0,0 +1,257 @@
...
+
+            // check that they are actually allowed to join the role
+            bool roleExistsInDb = CheckRoleEntryExists(role, Context.Guild);
+            if (roleExistsInDb == false)
+            {
+                await Context.Channel.SendMessageAsync("You cannot join that role.");
+                return;
+            }
+
+            // if they have the role we remove it
```
[the remark this answers is not in the document]
Comment (iShauny) [reply Floofbot/Modules/UserAssignableRoleCommands.cs:67]: This is misleading. It was an old comment from previous legacy code. It does not actually remove it

File: Floofbot/Modules/UserAssignableRoleCommands.cs (hunk 1 above)
Comment (iShauny) [reply Floofbot/Modules/UserAssignableRoleCommands.cs:45]: Set default value to null

File: Floofbot/Modules/UserAssignableRoleCommands.cs (hunk 2 above)
Comment (iShauny) [reply Floofbot/Modules/UserAssignableRoleCommands.cs:60]: yes all roles in the database are ones that users are allowed to join

File: Floofbot/Modules/UserAssignableRoleCommands.cs [hunk 13]
```diff
@@ -0,0 +1,260 @@
...
+        }
+    }
+
+    [Summary("Configures the joinable user roles")]
+    [RequireUserPermission(GuildPermission.Administrator)]
+    [Group("iamconfig")]
+    public class UserAssignableRoleAdminCommands : InteractiveBase
+    {
+        FloofDataContext _floofDb;
+        private readonly Discord.Color EMBED_COLOUR = new Discord.Color((uint)new Random().Next(0x1000000));
```
Comment (JasperGray) [inline Floofbot/Modules/UserAssignableRoleCommands.cs:153]: You might want to redefine GenerateColor() here

File: Floofbot/Modules/UserAssignableRoleCommands.cs [hunk 14]
```diff
@@ -0,0 +1,248 @@
...
+using Floofbot.Services.Repository;
+using Floofbot.Services.Repository.Models;
+using System.Linq;
+using Discord.Addons.Interactive;
+using Discord.WebSocket;
+using Serilog;
+
+namespace Floofbot.Modules
+{
+    public class UserAssignableRoleCommands : InteractiveBase
```
Comment (JasperGray) [inline Floofbot/Modules/UserAssignableRoleCommands.cs:15]: What's the module that shows up when doing .help?
I wonder if we need a summary for the overall module

**Pull request merged**